### ISSUE TITLE
Matrix issues on combined Linux/MacOs build

### ISSUE THREAD

**Issue by SlySven** [travis-ci/travis-ci#5116]
Um, I may have the wrong end of the stick here but trying to compile [Mudlet](https://github.com/Mudlet/Mudlet) (which is targeted at all of Linux/MacOs/Windows Os Platforms) from both qmake and cmake native project files I'm having difficulties with specifying compiler and allowing failures.  On Linux I'd like to use GCC but having Clang take a look could make for more informative failures. On MacOs I guess I'd like to use Clang because the GCC is the ancient _Apple-munged_ 4.2 and we've now using C+11 features that I think THAT compiler cannot even imagine.

I somewhat naïvely tried this to be the start of our ```.travis.yml``` and ran it through the [validator](http://lint.travis-ci.org/) and it confirm the total breakage that I'd experienced with a similar file in actual use.
```
language: cpp
os:
  - osx
  - linux
compiler:
  - gcc
  - clang
env:
  - Q_OR_C_MAKE=qmake
  - Q_OR_C_MAKE=cmake
matrix:
    allow_failures:
    - os: linux
        compiler: clang
          env: Q_OR_C_MAKE=cmake
    - os: macos
        compiler: gcc
```
Things bork at the compiler specification in the allow failures section and I anticipate similar problems if I tried to limit the build matrix to gcc for Linux and clang for MacOs.  Am I doing this wrong or is it a real issue: _**compiler** (and possibly **os**) cannot be included as elements of the matrix_ ?

Our project _is_ feature flagged for use of the MacOs platform.

_EDIT_
I managed to get something that was accepted by using the environmental variables that are set for the platform:
```
matrix:
    allow_failures:
    - env: CXX=g++ & TRAVIS_OS_NAME=linux & Q_OR_C_MAKE=cmake,
    - env: CXX=clang & TRAVIS_OS_NAME=macos & Q_OR_C_MAKE=qmake
```
But this did not trigger some parts of the build to be "allowed to fail" - I guess my attempt to AND combinations is not correct - so can Os and Compiler be factored into the matrix and how _is_ it done?

P.S. until recently I thought that Ruby was a semi-precious gem and Rake was something I used to gather up fallen leaves so if either of those have a factor in the syntax for the .travis.yml file I guess I've got some more reading to do. :confused:

**Comment by SlySven** [comment]
Have just found and read [this specification](http://www.yaml.org/spec/1.2/spec.html#id2759963) for YAML - I am going to go and lie down in a darkened room now, I may be some time... :scream:

**Comment by BanzaiMan** [comment]
Can this be closed, then?

**Comment by SlySven** [comment]
I would humble suggest the on-line help documentation might benefit from a little expansion in this area.  I do not (yet :pray:) understand how to combine environment variables (and I doubt I would be the only one).  Perhaps a little more fundamentally, it is not clear to me how to code up for different combinations of __compiler:__ / __env:__ for different values of __os:__ - is it even possible?

Can you give me another 24 hours to experiment (or straighten me - and anyone else trying to get their heads around this aspect - out) on what can go into the matrix; _"Take the Red Pill, Neo"_ ...

**Comment by BanzaiMan** [comment]
@SlySven We are happy to help, but I am afraid I do not quite understand what confused you. Your original YAML snippet:

```yaml
language: cpp
os:
  - osx
  - linux
compiler:
  - gcc
  - clang
env:
  - Q_OR_C_MAKE=qmake
  - Q_OR_C_MAKE=cmake
matrix:
    allow_failures:
    - os: linux
        compiler: clang
          env: Q_OR_C_MAKE=cmake
    - os: macos
        compiler: gcc
```

does not specify one of the `allow_failures` job the way I suspect you intended (due to a bad indentation for `env: Q_OR_C_MAKE=cmake`).

**Comment by SlySven** [comment]
Yes, I think I am starting to get it now, I should have used:
```
language: cpp
os:
  - osx
  - linux
compiler:
  - gcc
  - clang
env:
  - Q_OR_C_MAKE=qmake
  - Q_OR_C_MAKE=cmake
matrix:
    allow_failures:
    - os: linux
      compiler: clang
      env: Q_OR_C_MAKE=cmake
    - os: osx
      compiler: gcc
```
This allows the following failures:
* Linux build AND using CLang compiler AND Q_OR_C_MAKE=cmake
* MacOs build AND using GCC compiler {implicitly for either Q_OR_C_MAKE=cmake or Q_OR_C_MAKE=qmake cases!}

The build matrix has 2x2x2 = 8 entries, and three of them can fail.

I think I've got it now, but can I suggest that:
* the documentation around [Build Matrix](http://docs.travis-ci.com/user/customizing-the-build/#Build-Matrix) is strengthened to stress the significance of the "-" as the start of what I might call a "case" or "combination" and indentation that groups (logical "And") keys of different things with the same indentation within that "case". :open_mouth:
* more attention is drawn to the on-line validator in that same section of the documentation, at least; and perhaps a pointer to some recommended further reading for YAML for ~~dummies~~ newbies?

:bulb:  It would be nice if that validator could be extended to expand any matrix found and display the results - possibly it could use the same "parser" that is used within the guts of the platform? :bulb:

Within the validator the "your repository must be feature flagged for the os setting to be used" is flagged as an error - can you reword it to say that the notice can be disregarded if the user's repository is so flagged.  At one stage a proposed version I tested like that came up with a green information message that everything was _hunky-dory_ (my words, not a literal quote) which was confusing as I had rearranged things so badly that the __os:__ key was ineffective - whereas I really should have got the "error" message!
